OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs

```diff
@@ -161,32 +147,39 @@ void BuildUnit(IBot bot, string name)
 
 		ActorInfo ChooseRandomUnitToBuild(ProductionQueue queue)
 		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var unitsToBuild = Info.UnitsToBuild.ToArray();
+			if (unitsToBuild.Length == 0)
 				return null;
 
-			var unit = buildableThings.Random(world.LocalRandom);
-			return HasAdequateAirUnitReloadBuildings(unit) ? unit : null;
-		}
-
-		ActorInfo ChooseUnitToBuild(ProductionQueue queue)
-		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var buildableThings = queue.BuildableItems().Shuffle(world.LocalRandom).ToArray();
+			if (buildableThings.Length == 0)
 				return null;
 
-			var myUnits = player.World
-				.ActorsHavingTrait<IPositionable>()
-				.Where(a => a.Owner == player)
-				.Select(a => a.Info.Name).ToList();
+			var allUnits = world.Actors.Where(a => a.Owner == player && Info.UnitsToBuild.ContainsKey(a.Info.Name) && !a.IsDead && a.IsInWorld).ToArray();
+
+			ActorInfo desiredUnit = null;
+			var desiredError = int.MaxValue;
+			foreach (var unit in buildableThings)
+			{
+				if (!Info.UnitsToBuild.ContainsKey(unit.Name) || (Info.UnitDelays != null && Info.UnitDelays.ContainsKey(unit.Name) && Info.UnitDelays[unit.Name] > world.WorldTick))
```

Comment:
Do not first check to see if something is in the dictionary to after retrieve it. Use TryGetValue to do it in one pass.

Reply:
Isn't the first always true? `buildableThings` is `Info.UnitsToBuild` so it will always contain the unit name?

Reply:
So if a unit delay is set to have a value greater than world tick, then skip that unit. Why?

Reply:
> Isn't the first always true? buildableThings is Info.UnitsToBuild so it will always contain the unit name?

`buildableThings` is from `queue.BuildableItems()`

> So if a unit delay is set to have a value greater than world tick, then skip that unit. Why?

This is the original design.